urnai/sc2/states/terran_state.py

```diff
@@ -0,0 +1,121 @@
+        new_state.append(get_my_units_amount(obs, units.Terran.MissileTurret) / 4)
+        new_state.append(get_my_units_amount(obs, units.Terran.SensorTower)/1)
+        new_state.append(get_my_units_amount(obs, units.Terran.Bunker)/4)
+        new_state.append(get_my_units_amount(obs, units.Terran.FusionCore))
+        new_state.append(get_my_units_amount(obs, units.Terran.GhostAcademy))
+        new_state.append(get_my_units_amount(obs, units.Terran.Barracks) / 3)
+        new_state.append(get_my_units_amount(obs, units.Terran.Factory) / 2)
+        new_state.append(get_my_units_amount(obs, units.Terran.Starport) / 2)
+
+        # Instead of making a vector for all coordnates on the map, we'll
+        # discretize our enemy space
+        # and use a 4x4 grid to store enemy positions by marking a square as 1 if
+        # there's any enemy on it.
+
+        enemy_grid = np.zeros((self.grid_size, self.grid_size))
+        player_grid = np.zeros((self.grid_size, self.grid_size))
+
+        enemy_units = [unit for unit in obs.raw_units if
+                       unit.alliance == features.PlayerRelative.ENEMY]
+        player_units = [unit for unit in obs.raw_units if
+                        unit.alliance == features.PlayerRelative.SELF]
+
+        for i in range(0, len(enemy_units)):
+            y = int(math.ceil((enemy_units[i].x + 1) / 64 / self.grid_size))
+            x = int(math.ceil((enemy_units[i].y + 1) / 64 / self.grid_size))
+            enemy_grid[x - 1][y - 1] += 1
+
+        for i in range(0, len(player_units)):
+            y = int(math.ceil((player_units[i].x + 1) / (64 / self.grid_size)))
+            x = int(math.ceil((player_units[i].y + 1) / (64 / self.grid_size)))
+            player_grid[x - 1][y - 1] += 1
+
+        if not self.base_top_left:
+            enemy_grid = np.rot90(enemy_grid, 2)
+            player_grid = np.rot90(player_grid, 2)
+
+        # Normalizing the values to always be between 0 and 1
+        # (since the max amount of units in SC2 is 200)
+        enemy_grid = enemy_grid / 200
+        player_grid = player_grid / 200
+
+        new_state.extend(enemy_grid.flatten())
+        new_state.extend(player_grid.flatten())
+        final_state = np.expand_dims(new_state, axis=0)
+
+        self._state = final_state
+        return final_state
+
+    @property
+    def state(self):
+        return self._state
+
+    @property
+    def dimension(self):
+        return self._state_size
+
+    def reset(self):
+        self._state = None
+        self.base_top_left = None
+
+def get_my_units_amount(obs, unit_type):
+    return len(get_units_by_type(obs, unit_type, features.PlayerRelative.SELF))
+
+def get_units_by_type(obs, unit_type, alliance=features.PlayerRelative.SELF):
+    return [unit for unit in obs.raw_units
+            if unit.unit_type == unit_type
+            and unit.alliance == alliance
+            and unit.build_progress == 100]
```

Comment:
These functions should be kept in another file so that they can be reused by other classes.

Reply:
Should this file be inside the same folder as the `TerranState` class ?

Reply:
For now, I think so. If they are used by other classes that are not on the same path, then we can look at another place to put them.

Reply:
About the unit tests, should a different test file be created for these methods?

Reply:
Ideally, the unit test should be in a path similar to the file it is testing, so yes.